### PR TITLE
Transfer - fix transfer delayed artifacts while having errors

### PR DESCRIPTION
**Pull request by attiasas**
- [x] All [tests](https://github.com/jfrog/jfrog-cli-core#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [x] All [static analysis checks](https://github.com/jfrog/jfrog-cli-core/actions/workflows/analysis.yml) passed.
- [x] This pull request is on the dev branch.
- [x] I used gofmt for formatting the code before submitting the pull request.
-----

### CODE REVIEW

File: artifactory/commands/transferfiles/delayedartifactshandler.go [hunk 1]
```diff
@@ -123,12 +182,63 @@ func consumeDelayedArtifactsFiles(pcWrapper *producerConsumerWrapper, filesToCon
...
+	}
+	return delayedArtifactsFile, nil
+}
+
+// Gets a list of all delay files from the CLI's cache for a specific repo
+func getDelayFiles(repoKeys []string) (filesPaths []string, err error) {
+
+	dirPath, err := coreutils.GetJfrogTransferDelaysDir()
+	if err != nil {
+		return []string{}, err
```
Comment (eyalbe4) [inline artifactory/commands/transferfiles/delayedartifactshandler.go:212]: return []string{}, err --> return

The same goes for the other retuern statements in this function.

File: artifactory/commands/transferfiles/delayedartifactshandler.go [hunk 2]
```diff
@@ -123,12 +182,63 @@ func consumeDelayedArtifactsFiles(pcWrapper *producerConsumerWrapper, filesToCon
...
+	err = json.Unmarshal(fContent, &delayedArtifactsFile)
+	if err != nil {
+		return delayedArtifactsFile, errorutils.CheckError(err)
+	}
+	return delayedArtifactsFile, nil
+}
+
+// Gets a list of all delay files from the CLI's cache for a specific repo
+func getDelayFiles(repoKeys []string) (filesPaths []string, err error) {
+
```
Comment (eyalbe4) [inline artifactory/commands/transferfiles/delayedartifactshandler.go:209]: Redundant new-line.

File: artifactory/commands/transferfiles/delayedartifactshandler.go [hunk 3]
```diff
@@ -216,3 +326,70 @@ func createdDelayedArtifactsChannelMng() DelayedArtifactsChannelMng {
 	channel := make(chan FileRepresentation, fileWritersChannelSize)
 	return DelayedArtifactsChannelMng{channel: channel}
 }
+
+type SplitContentWriter struct {
```
Comment (eyalbe4) [inline artifactory/commands/transferfiles/delayedartifactshandler.go:330]: Let's add a description for this struct. Explain why it is needed.
Let's also describe all of the struct's components.
Also, it looks like this struct doesn't have have to be public.

File: artifactory/commands/transferfiles/delayedartifactshandler.go [hunk 4]
```diff
@@ -216,3 +326,70 @@ func createdDelayedArtifactsChannelMng() DelayedArtifactsChannelMng {
...
+	writer *content.ContentWriter
+
+	arrayKey       string
+	maxRecordAllow int
+	dirPath        string
+	filePrefix     string
+
+	recordCount  int
+	fileIndex    int
+	ContentFiles []string
```
Comment (eyalbe4) [inline artifactory/commands/transferfiles/delayedartifactshandler.go:340]: Why does this property need to be public?

File: artifactory/commands/transferfiles/delayedartifactshandler.go [hunk 5]
```diff
@@ -216,3 +326,70 @@ func createdDelayedArtifactsChannelMng() DelayedArtifactsChannelMng {
...
+}
+
+func NewSplitContentWriter(key string, maxRecordsPerFile int, directoryPath string, prefix string) *SplitContentWriter {
```
Comment (eyalbe4) [inline artifactory/commands/transferfiles/delayedartifactshandler.go:343]: This function probably doesn't need to be public.

File: artifactory/commands/transferfiles/delayedartifactshandler.go [hunk 6]
```diff
@@ -216,3 +326,70 @@ func createdDelayedArtifactsChannelMng() DelayedArtifactsChannelMng {
...
+	scw := SplitContentWriter{arrayKey: key, maxRecordAllow: maxRecordsPerFile, dirPath: directoryPath, filePrefix: prefix, ContentFiles: []string{}}
+	return &scw
+}
+
+// Create new file if needed, writes a record and closes a file if it reached its maxRecord
+func (w *SplitContentWriter) WriteRecord(record interface{}) error {
```
Comment (eyalbe4) [inline artifactory/commands/transferfiles/delayedartifactshandler.go:349]: It looks like this function doesn't have to be public.

File: artifactory/commands/transferfiles/delayedartifactshandler_test.go [hunk 7]
```diff
@@ -50,8 +60,31 @@ func TestDelayedArtifactsMng(t *testing.T) {
...
+	// add not relevant files to confuse
+	for i := 0; i < 4; i++ {
+
+		assert.NoError(t, ioutil.WriteFile(filepath.Join(delaysDirPath, getDelayMockFileName(delayTestRepoKey, i)), nil, 0644))
+	}
+	assert.NoError(t, ioutil.WriteFile(filepath.Join(delaysDirPath, getDelayMockFileName("wrong-"+testRepoKey, 1)), nil, 0644))
+	assert.NoError(t, ioutil.WriteFile(filepath.Join(delaysDirPath, getDelayMockFileName(testRepoKey+"-wrong", 0)), nil, 0644))
+	assert.NoError(t, ioutil.WriteFile(filepath.Join(delaysDirPath, getDelayMockFileName(testRepoKey+"-0-0", 0)), nil, 0644))
+	assert.NoError(t, ioutil.WriteFile(filepath.Join(delaysDirPath, getDelayMockFileName(testRepoKey+"-1", 0)), nil, 0644))
+	assert.NoError(t, ioutil.WriteFile(filepath.Join(delaysDirPath, getDelayMockFileName("wrong-"+testRepoKey+"-wrong", 0)), nil, 0644))
```
Comment (yahavi) [inline artifactory/commands/transferfiles/delayedartifactshandler_test.go:73]: This line repeats 6 times:
```go
assert.NoError(t, ioutil.WriteFile(filepath.Join(delaysDirPath, getDelayMockFileName("xxx", 111)), nil, 0644))
```
Can we add a loop or use a dedicated function to reduce code duplication?

File: artifactory/commands/transferfiles/fileserror.go [hunk 8]
```diff
@@ -140,7 +141,18 @@ func (e *errorsRetryPhase) initProgressBar() error {
...
 
-	e.progressBar.AddPhase3(int64(filesCount))
+	// The progress bar will also be responsible to display the number of delayed items for this repository.
+	// Those delayed artifacts will be handled at the end of this phase in case they exist.
+	delayCount := 0
+	delayFiles, err := getDelayFiles([]string{e.repoKey})
+	if err != nil {
+		return err
+	}
+	delayCount, err = countDelayFilesContent(delayFiles)
```
Comment (yahavi) [inline artifactory/commands/transferfiles/fileserror.go:151]: A bit shorter:
```suggestion
	delayFiles, err := getDelayFiles([]string{e.repoKey})
	if err != nil {
		return err
	}
	delayCount, err := countDelayFilesContent(delayFiles)
```

File: artifactory/commands/transferfiles/delayedartifactshandler.go [hunk 9]
```diff
@@ -216,3 +331,77 @@ func createdDelayedArtifactsChannelMng() DelayedArtifactsChannelMng {
...
 }
+
+// SplitContentWriter writes to files a single JSON object that holds a list of records added as stream.
+// It can limit the amount of records per file and splits the content to several files if needed.
+type SplitContentWriter struct {
+	writer *content.ContentWriter
+	// JSON array key of the object
+	arrayKey string
+	// Limit for the amount of records allowed per file
+	maxRecordAllow int
```
Comment (yahavi) [inline artifactory/commands/transferfiles/delayedartifactshandler.go:342]:
```suggestion
	maxRecordsAllowed int
```

File: artifactory/commands/transferfiles/delayedartifactshandler.go [hunk 10]
```diff
@@ -216,3 +331,77 @@ func createdDelayedArtifactsChannelMng() DelayedArtifactsChannelMng {
...
+	// The path for the directory that will hold the files of the content
+	dirPath string
+	// The name for the files that will be generated (a counter will added as a suffix to the files by this writer)
+	filePrefix string
+	// Counter for the amount of records at the current file
+	recordCount int
+	// Counter for amount if files generated for the content
+	fileIndex int
+	// List all the file paths of the files that were generated for the content
```
Comment (yahavi) [inline artifactory/commands/transferfiles/delayedartifactshandler.go:351]:
```suggestion
	// List of all the paths of the files that were generated for the content
```

File: artifactory/commands/transferfiles/delayedartifactshandler.go [hunk 11]
```diff
@@ -216,3 +331,77 @@ func createdDelayedArtifactsChannelMng() DelayedArtifactsChannelMng {
...
+	contentFiles []string
+}
+
+func newSplitContentWriter(key string, maxRecordsPerFile int, directoryPath string, prefix string) *SplitContentWriter {
+	scw := SplitContentWriter{arrayKey: key, maxRecordAllow: maxRecordsPerFile, dirPath: directoryPath, filePrefix: prefix, contentFiles: []string{}}
+	return &scw
```
Comment (yahavi) [inline artifactory/commands/transferfiles/delayedartifactshandler.go:357]:
```suggestion
	return &SplitContentWriter{arrayKey: key, maxRecordAllow: maxRecordsPerFile, dirPath: directoryPath, filePrefix: prefix}
```

**Review by yahavi — CHANGES_REQUESTED** [review]
Looks good, @attiasas!

**Review by yahavi — APPROVED** [review]

**Pull request merged**